spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,566 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.sharing.spark
+
+// scalastyle:off import.ordering.noEmptyLine
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.{DataFrame, DeltaSharingScanUtils, SparkSession}
+
+import org.apache.spark.sql.connector.read.streaming
+import org.apache.spark.sql.connector.read.streaming.{
+  ReadAllAvailable,
+  ReadLimit,
+  ReadMaxFiles,
+  SupportsAdmissionControl
+}
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+import org.apache.spark.sql.execution.streaming._
+import org.apache.spark.sql.types.StructType
+
+import io.delta.sharing.spark.model.{AddFile, AddFileForCDF, DeltaTableFiles}
+
+/**
+ * A case class to help with `Dataset` operations regarding Offset indexing, representing AddFile
+ * actions in a Delta log.
+ * For proper offset tracking(move the offset to the next version if all data is consumed in the
+ * current version), there are also special sentinel values with index = -1 and add = null.
+ *
+ * This class is not designed to be persisted in offset logs or such.
+ *
+ * @param version The version of the Delta log containing this AddFile.
+ * @param index The index of this AddFile in the Delta log.
+ * @param add The AddFile.
+ * @param isLast A flag to indicate whether this is the last AddFile in the version. This is used
+ *               to resolve an off-by-one issue in the streaming offset interface; once we've read
+ *               to the end of a log version file, we check this flag to advance immediately to the
+ *               next one in the persisted offset. Without this special case we would re-read the
+ *               already completed log file.
+ */
+private[sharing] case class IndexedFile(
+  version: Long,
+  index: Long,
+  add: AddFile,
+  isLast: Boolean = false) {
+
+  def getFileAction: AddFile = { add }
+}
+
+/**
+ * Base trait for the Delta Sharing Source, that contains methods that deal with
+ * getting changes from the delta sharing server.
+ * TODO(lin.zhou) Support SupportsTriggerAvailableNow
+ */
+trait DeltaSharingSourceBase extends Source
+  with SupportsAdmissionControl
+  with Logging { self: DeltaSharingSource =>
```

Comment:
Same question: What is the use case for this design pattern where we use a trait with the self directive? What benefit does it provide?